Compile-time comparison of optionals (== / !=)

theInkSquid opened the pull request:
This will allow code like this to function at compile time, partially implementing #1332:

```rust
const x: ?i32 = get_some_optional();
const y: ?i32 = get_another_optional();
if (x == y) {
    do_a_thing();
}

if (x != y) {
    do_another_thing();
}
```
Which currently only works for optional pointers.
I will work on making this runtime soon. 

One thing I am not completely sure about: when calling `ir_analyze_optional_value_payload_value()`, I am not sure if it should be given `op1/op2->base`, or `bin_op_instruction->base.base` for the `source_instr` parameter.

mikdusan commented:
> One thing I am not completely sure about: when calling `ir_analyze_optional_value_payload_value()`, I am not sure if it should be given `op1/op2->base`, or `bin_op_instruction->base.base` for the `source_instr` parameter.

#### `foo.zig`
```zig
export fn entry() void {
    var oa: ?u32 = 999;
    var a = oa orelse 0;
    if (oa) |uw| {}
}
```

`ir_analyze_optional_value_payload_value` seems to be a pretty rare call, so trapping for `ir_analyze_unwrap_optional_payload` instead the above source code hits 3 times:

source_instr | line:col | base_ptr | line:col
--- | --- | --- | ---
`999` | 2:20 | `var` | 2:5
`orelse` | 3:16 | `var` | 2:5
`if` | 4:15 | `var` | 2:5

so I'd say `bin_op_instruction` is good for `source_instr`

theInkSquid commented:
@mikdusan Alright, I fixed it. I also added error checking after the unwraps, which after looking in the other code that uses the function, seems necessary.

theInkSquid commented:
This did not have intended behavior. I will be back soon with a correct version also with runtime support.
